### PR TITLE
Add texture support from CuTextures.jl

### PR DESCRIPTION
**Pull request by maleadt**
I messed up and https://github.com/JuliaGPU/CUDA.jl/pull/206 got closed.

### CODE REVIEW

File: lib/cuda/texture.jl [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+Note that it is necessary the their memory is well aligned and strided (good pitch).
+Currently, that is not being enforced.
+"""
+CuTexture(x::CuArray{T,N}; kwargs...) where {T,N} =
+    CuTexture{T,N,1}(x; kwargs...)
+CuTexture(x::CuArray{NTuple{C,T},N}; kwargs...) where {T,N,C} =
+    CuTexture{T,N,C}(x; kwargs...)
+
+Adapt.adapt_storage(::Adaptor, t::CuTexture{T,N,C}) where {T,N,C} =
+    CuDeviceTexture{T,N,C,t.normalized_coordinates}(size(t), t.handle)
```
Comment (cdsousa) [inline lib/cuda/texture.jl]: (copying comment from https://github.com/JuliaGPU/CUDA.jl/pull/206)
Can this bring any type of instability?
I have not used a solution like this in the first place due to the fear of type instability, but now I guess that maybe the kernel launch works as a function barrier, is that right?

Comment (maleadt) [reply lib/cuda/texture.jl]: Yes, it works like a function barrier indeed, we never get type instabilities across the host<->device barrier (unless we legitimately were dealing with an abstract type.

**Comment by maleadt** [comment]
I wonder if it would be valuable to try and expose this through array abstractions like broadcast. Since those typically require matching sizes, I could imagine something like `warp(::CuTexture, dims...)` creating a 'view' with changed dimensions so that we can have it interpolate. Or maybe `warp(f, ::CuTexture)` to apply a mapping to indices and, e.g., be able to do the sine warp example.

@chrisrackauckas, you've expressed interest in this, what's your use case?

**Comment by maleadt** [comment]
```cuda
julia> a = CUDA.rand(2,2)
2×2 CuArray{Float32,2,Nothing}:
 0.148875   0.386196
 0.0170659  0.696161

julia> b = similar(a)
2×2 CuArray{Float32,2,Nothing}:
 0.0  0.0
 0.0  0.0

julia> b .= CuTexture(a)
2×2 CuArray{Float32,2,Nothing}:
 0.148875   0.148875
 0.0170659  0.0170659

julia> @device_code_llvm debuginfo=:none b .= CuTexture(a)
; PTX CompilerJob of kernel broadcast(CUDA.CuKernelContext, CuDeviceArray{Float32,2,CUDA.AS.Global}, Base.Broadcast.Broadcasted{Nothing,Tuple{Base.OneTo{Int64},Base.OneTo{Int64}},typeof(identity),Tuple{Base.Broadcast.Extruded{CuDeviceTexture{Float32,2,false},Tuple{Bool,Bool},Tuple{Int64,Int64}}}}) for sm_75

...
L152:                                             ; preds = %L146, %L141
  %31 = load i64, i64* %56, align 8
  %32 = call [4 x float] @llvm.nvvm.tex.unified.2d.v4f32.s32(i64 %31, i32 %57, i32 %28)
  %.fca.0.extract = extractvalue [4 x float] %32, 0
  %33 = getelementptr inbounds { [2 x i64], i64 }, { [2 x i64], i64 }* %0, i64 0, i32 1
  %34 = bitcast i64* %33 to float addrspace(1)**
  %35 = load float addrspace(1)*, float addrspace(1)** %34, align 8
  %36 = getelementptr float, float addrspace(1)* %35, i64 %27
  store float %.fca.0.extract, float addrspace(1)* %36, align 4
  br label %L33
...
}
```

**Comment by maleadt** [comment]
CI is green, so let's merge this. API is not final yet, it requires explicit use of CuTexture/CuTextureArray objects (more or less like @cdsousa designed it, but integrated with the rest of the stack). I'd like to try making CuArray compatible with it, dispatching on the inner buffer object.

**Comment by cdsousa** [comment]
Nice!!!

The issue above, got fixed? I've started looking at it and was able to test in my original code, where it was ok, but then I had no more time to continue.

**Comment by maleadt** [comment]
> The issue above, got fixed? I've started looking at it and was able to test in my original code, where it was ok, but then I had no more time to continue.

Which issue? The tests are mostly the same as they were in your version.

**Comment by cdsousa** [comment]
> > The issue above, got fixed? I've started looking at it and was able to test in my original code, where it was ok, but then I had no more time to continue.

https://github.com/JuliaGPU/CUDA.jl/pull/209#issuecomment-641279792

Wasn't broadcast assignment supposed to give the same as a copy?

**Comment by maleadt** [comment]
Ah, I didn't even see they mismatched! I was just happy it generated appropriate code. But it seems to work now:

```
julia> a = CUDA.rand(8,8)
8×8 CuArray{Float32,2,Nothing}:
 0.916239  0.976847   0.755844  0.66619   0.011888  0.213493   0.665336   0.132537
 0.547251  0.867846   0.316787  0.580942  0.953204  0.762209   0.0874605  0.0591269
 0.413512  0.374084   0.369481  0.642127  0.724004  0.0811296  0.657992   0.788099
 0.48726   0.417901   0.889191  0.795766  0.334781  0.387315   0.913375   0.881644
 0.943592  0.584336   0.455078  0.28139   0.591649  0.0605415  0.832624   0.409376
 0.57543   0.0464874  0.612669  0.451585  0.950189  0.647614   0.658178   0.89397
 0.987634  0.586837   0.71754   0.884229  0.501966  0.217039   0.520088   0.542805
 0.730895  0.838332   0.796267  0.295658  0.348608  0.727785   0.867374   0.255682

julia> b = similar(a)

julia> b .= CuTexture(a)
8×8 CuArray{Float32,2,Nothing}:
 0.916239  0.976847   0.755844  0.66619   0.011888  0.213493   0.665336   0.132537
 0.547251  0.867846   0.316787  0.580942  0.953204  0.762209   0.0874605  0.0591269
 0.413512  0.374084   0.369481  0.642127  0.724004  0.0811296  0.657992   0.788099
 0.48726   0.417901   0.889191  0.795766  0.334781  0.387315   0.913375   0.881644
 0.943592  0.584336   0.455078  0.28139   0.591649  0.0605415  0.832624   0.409376
 0.57543   0.0464874  0.612669  0.451585  0.950189  0.647614   0.658178   0.89397
 0.987634  0.586837   0.71754   0.884229  0.501966  0.217039   0.520088   0.542805
 0.730895  0.838332   0.796267  0.295658  0.348608  0.727785   0.867374   0.255682

julia> a == b
true
```

Although the exact example doesn't anymore:

```
julia> CuTexture(CUDA.rand(2,2))
ERROR: CUDA error: invalid argument (code 1, ERROR_INVALID_VALUE)
```

Anyway, CuTexture isn't really ready to be used with broadcast yet, but it's nice to see the essentials work already.

**Pull request merged**